File: job-scraper-backend/scrapers/opphubafrica_scraper.py
from .base_scraper import BaseScraper
import requests
from bs4 import BeautifulSoup
from datetime import datetime, timedelta
import re
from .keywords import DEFAULT_KEYWORDS

import json

class OppHubAfricaScraper(BaseScraper):
    """A scraper for 'opphubafrica.com' job listings."""

    def _parse_relative_date(self, date_str):
        """
        Parses relative date strings like '3 days from now', '1 week ago', etc.
        Returns a datetime object or None if parsing fails.
        """
        if not date_str:
            return None
            
        today = datetime.now()
        date_str = date_str.lower().strip()
        
        if date_str == 'tomorrow':
            return today + timedelta(days=1)
        if date_str == 'yesterday':
            return today - timedelta(days=1)
        if date_str == 'today':
            return today

        # Regex to find number and unit
        match = re.search(r'(\d+)\s+(day|week|month|year)s?\s+(from now|ago)', date_str)
        if match:
            number = int(match.group(1))
            unit = match.group(2)
            direction = match.group(3)
            
            delta = timedelta()
            if unit == 'day':
                delta = timedelta(days=number)
            elif unit == 'week':
                delta = timedelta(weeks=number)
            elif unit == 'month':
                # Approximation
                delta = timedelta(days=number * 30)
            elif unit == 'year':
                delta = timedelta(days=number * 365)
            
            if direction == 'ago':
                return today - delta
            else:
                return today + delta
                
        return None

    def scrape(self, keyword=None):
        BASE_URL = "https://opphubafrica.com"
        HEADERS = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        }
        
        print(f"Scraping {BASE_URL}...")
        try:
            response = requests.get(BASE_URL, headers=HEADERS, timeout=20)
            response.raise_for_status()
            html_content = response.text
        except Exception as e:
            print(f"Error fetching {BASE_URL}: {e}")
            return {"total_jobs": 0, "unique_companies": 0, "jobs": []}

        all_jobs = []
        company_names = set()
        
        try:
            # Extract the 'opportunities' JSON array from the embedded script
            # Look for: opportunities: [...]
            # We use re.DOTALL to handle multiline, but based on the file inspection it might be minimized/inline.
            # We'll try to match the specific variable assignment in the Alpine data function.
            match = re.search(r'opportunities:\s*(\[.*?\])\s*(?:,?\s*\n|,\s*page)', html_content, re.DOTALL)
            
            if not match:
                print("Could not find inline opportunities variable in HTML source.")
                # Fallback: maybe the server renders it differently?
                # One more try with simpler boundary
                match = re.search(r'opportunities:\s*(\[{.*}\]),', html_content)
            
            if match:
                raw_json = match.group(1)
                opportunities = json.loads(raw_json)
                print(f"Found {len(opportunities)} raw opportunities")

                for item in opportunities:
                    # 1. Check type
                    if item.get("type") != "job":
                        continue

                    job_id = item.get("id")
                    title = item.get("title")
                    company = item.get("company_name", "Unknown")
                    # Handling relative URL provided in JSON or constructing it
                    link = item.get("url")
                    if not link:
                         slug = item.get("slug")
                         if slug:
                            link = f"{BASE_URL}/jobs/{slug}"
                         elif job_id:
                            link = f"{BASE_URL}/jobs/{job_id}" # Fallback
                    
                    location = item.get("location", "Unknown")
                    
                    deadline_str = item.get("deadline", "N/A")
                    deadline_date = None
                    parsed_date = self._parse_relative_date(deadline_str)
                    
                    if parsed_date:
                        deadline_date = parsed_date.strftime("%Y-%m-%d")
                    else:
                        deadline_date = deadline_str # Fallback

                    job_data = {
                        "title": title,
                        "company": company,
                        "link": link,
                        "location": location,
                        "deadline_date": deadline_date if deadline_date else "N/A",
                    }
                    
                    all_jobs.append(job_data)
                    company_names.add(company)
            else:
                print("Regex match failed for opportunities data.")

        except Exception as e:
            print(f"Error parsing embedded JSON: {e}")

        # --- Filtering ---
        filtered_jobs = []
        if keyword:
            pattern = r'\b' + re.escape(keyword) + r'\b'
            for job in all_jobs:
                if re.search(pattern, job['title'], re.IGNORECASE):
                    filtered_jobs.append(job)
        else:
            # Default IT filtering
            pattern = re.compile(r'\b(' + '|'.join(DEFAULT_KEYWORDS) + r')\b', re.IGNORECASE)
            for job in all_jobs:
                title = job.get('title', '')
                if title and pattern.search(title):
                    filtered_jobs.append(job)

        return {
            "total_jobs": len(all_jobs),
            "unique_companies": len(company_names),
            "jobs": filtered_jobs
        }

File: job-scraper-backend/src/scrapers/oppHubAfricaScraper.js
// src/scrapers/oppHubAfricaScraper.js

import axios from 'axios';
import * as cheerio from 'cheerio';
import { BaseScraper } from './baseScraper.js';

export class OppHubAfricaScraper extends BaseScraper {
  /**
   * Scraper for 'opphubafrica.com' job listings.
   * Extracts inline JSON data from the HTML page.
   */

  _parseRelativeDate(dateStr) {
    if (!dateStr) return null;

    const today = new Date();
    const str = dateStr.toLowerCase().trim();

    if (str === 'tomorrow') {
      return new Date(today.getTime() + 24 * 60 * 60 * 1000);
    }
    if (str === 'yesterday') {
      return new Date(today.getTime() - 24 * 60 * 60 * 1000);
    }
    if (str === 'today') {
      return new Date(today);
    }

    // Match patterns like "3 days from now", "1 week ago"
    const match = str.match(/(\d+)\s+(day|week|month|year)s?\s+(from now|ago)/);
    if (match) {
      const number = parseInt(match[1]);
      const unit = match[2];
      const direction = match[3];

      let deltaMs = 0;
      switch (unit) {
        case 'day':
          deltaMs = number * 24 * 60 * 60 * 1000;
          break;
        case 'week':
          deltaMs = number * 7 * 24 * 60 * 60 * 1000;
          break;
        case 'month':
          deltaMs = number * 30 * 24 * 60 * 60 * 1000;
          break;
        case 'year':
          deltaMs = number * 365 * 24 * 60 * 60 * 1000;
          break;
      }

      if (direction === 'ago') {
        return new Date(today.getTime() - deltaMs);
      } else {
        return new Date(today.getTime() + deltaMs);
      }
    }

    return null;
  }

  async fetchAll() {
    // The main jobs page where listings are displayed
    const JOBS_URL = 'https://opphubafrica.com/jobs';
    const BASE_URL = 'https://opphubafrica.com';
    const HEADERS = {
      'User-Agent':
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
    };

    console.log(`Scraping ${JOBS_URL}...`);

    let htmlContent;
    try {
      const response = await axios.get(JOBS_URL, { headers: HEADERS, timeout: 20000 });
      htmlContent = response.data;
    } catch (err) {
      console.error(`Error fetching ${JOBS_URL}: ${err.message}`);
      return { allJobs: [], companyNames: [] };
    }

    const allJobs = [];
    const companyNames = new Set();

    try {
      const $ = cheerio.load(htmlContent);
      const articles = $('article');
      console.log(`Found ${articles.length} job articles`);

      articles.each((i, el) => {
        const titleEl = $(el).find('h2, h3, h4').first();
        const title = titleEl.text().trim();
        const companyEl = titleEl.next();
        const company = companyEl.length ? companyEl.text().trim() : 'Unknown';
        
        let linkPath = $(el).find('a').last().attr('href') || '';
        let link = linkPath;
        if (linkPath && !linkPath.startsWith('http')) {
           link = `${BASE_URL}${linkPath}`;
        }

        const spans = [];
        $(el).find('span').each((j, span) => {
           const t = $(span).text().trim();
           if(t) spans.push(t);
        });
        
        const deadlineStr = spans.length > 0 ? spans[0] : '';
        const location = spans.length > 1 ? spans[spans.length - 1] : 'Unknown';

        let deadlineDate;
        if (deadlineStr) {
          const parsedDate = this._parseRelativeDate(deadlineStr);
          if (parsedDate) {
            deadlineDate = parsedDate.toISOString().split('T')[0];
          } else {
            deadlineDate = deadlineStr;
          }
        } else {
          deadlineDate = 'N/A';
        }

        if (title) {
          allJobs.push({
            title,
            company,
            link,
            location,
            published_date: deadlineDate,
          });
          companyNames.add(company);
        }
      });
    } catch (err) {
      console.error(`Error parsing HTML: ${err.message}`);
    }

    // --- Return raw data for caching ---
    return { allJobs, companyNames: [...companyNames] };
  }
}
